### PR TITLE
Bug 1919271: [on-prem] NM resolve prepender: support appending a nameserver for systemd-resolved

### PR DESCRIPTION
**Pull request by vrutkovs**
FCOS 33 doesn't use NM to manage resolv.conf. Instead it uses systemd-resolved.

The script should not replace /etc/resolv.conf in this case, but instead create a dropin for systemd-resolved

4.6 cherry-pick: https://github.com/openshift/machine-config-operator/pull/2356

### CODE REVIEW

**Review by yuqi-zhang — APPROVED** [review]
Code looks fine, I'm not sure if the `Only leave the first 3 nameservers in /etc/resolv.conf` moving into the `$NAMESERVER_IP` condition affects anything. It'd be good for the metal team to have the final lgtm

File: templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml
Comment (darkmuggle) [inline templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml:57]: https://www.freedesktop.org/software/systemd/man/resolved.conf.html

I *think* we should use something like `60-kni.conf` to make ordering easier.

Comment (vrutkovs) [reply templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml]: Good idea, updated

**Comment by vrutkovs** [comment]
>I'm not sure if the Only leave the first 3 nameservers in /etc/resolv.conf moving into the $NAMESERVER_IP condition affects anything

Nameservers would be cleanup on RHCOS when new interface is up and NAMESERVER_IP is resolved (previously we'd clean those regardless of NAMESERVER_IP result). I don't think this would regress in any corner case really

**Comment by kikisdeliveryservice** [comment]
holding for on-prem reviews

/hold

File: templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml
Comment (mandre) [inline templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml:56]: What sets the $NAME variable? Don't you also need to check the version of Fedora?

Perhaps a better check would be to look for the use of systemd-resolved, so that there's no surprise when RHCOS switches to it.

Comment (vrutkovs) [reply templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml:56]: > What sets the $NAME variable?

Its being set after `source /etc/os-release`

> Don't you also need to check the version of Fedora?

We could add this later, but OKD 4.6 uses F33+, so systemd-resolved would certainly be used

> Perhaps a better check would be to look for the use of systemd-resolved

That was my initial approach, but its not clear if there is a nice way to do that - check that /etc/resolv.conf is a symlink? systemd-resolved service is running?

**Comment by iamemilio** [comment]
/lgtm 

I think this is good to go

File: templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml
Comment (mandre) [reply templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml:56]: I think the most robust is `systemctl is-active --quiet systemd-resolved`

**Review by Fedosin — APPROVED** [review]
/lgtm

File: templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml
Comment (vrutkovs) [reply templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml:56]: For some reason that didn't work. I'll have another look and follow up on that later

**Review by Fedosin — APPROVED** [review]
/lgtm

File: templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml
Comment (mandre) [reply templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml:56]: Yeah, don't worry too much about it. I think we'll know where to look if suddenly DNS stops working.

**Review by mandre — COMMENTED** [review]
/lgtm

**Comment by kikisdeliveryservice** [comment]
I see a bunch of openstack approvals,

@bcrochet @celebdor @jcpowermac PTAL

/test e2e-vsphere

**Comment by kikisdeliveryservice** [comment]
(pinged in slack channels)

File: templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml
Comment (celebdor) [reply templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml:56]: I think checking for the link should be fine.

File: templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml
Comment (celebdor) [inline templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml:59]: Wouldn't it be nicer to have a single heredoc or an echo with "\n" instead of three separate echo?

Comment (vrutkovs) [reply templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml:59]: The most cleanest way would be creating this file unconditionally - and it would be caught up in RHEL when it does the switch. WDYT?

Comment (vrutkovs) [reply templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml:59]: I don't have a preference how to write the file, seems echo with '\n' would the cleanest option. I don't mind rewriting it if you insist

**Comment by cybertron** [comment]
/lgtm

Seems like everyone is in agreement on the functionality of this. If we want to modify how things get done we can always do that in a followup.

**Comment by vrutkovs** [comment]
Skipping optional tests
/skip

**Pull request merged**
